azhidkov.pro/content/posts/21/04/en_why-jpa-should-be-avoided.adoc

```diff
@@ -0,0 +1,359 @@
+. It's a 'reliable solution from a trusted vendor,' which makes it easier to sell to your client or CTO than other alternatives.
+
+Nonetheless, SDJ's design philosophy is quite programmer-friendly:
+[quote, Spring Data JDBC Reference Documentation, https://docs.spring.io/spring-data/jdbc/docs/2.1.7/reference/html/#jdbc.why]
+____
+Spring Data JDBC aims to be much simpler conceptually, by embracing the following design decisions:
+
+* If you load an entity, SQL statements get run. Once this is done, you have a completely loaded entity. No lazy loading or caching is done.
+
+* If you save an entity, it gets saved. If you do not, it does not. There is no dirty tracking and no session.
+
+* There is a simple model of how to map entities to tables. It probably only works for rather simple cases. If you do not like that, you should code your own strategy. Spring Data JDBC offers only very limited support for customizing the strategy with annotations.
+____
+
+A bit down below we read:
+[quote, Spring Data JDBC Reference Documentation, https://docs.spring.io/spring-data/jdbc/docs/2.1.7/reference/html/#mapping.general-recommendations]
+____
+
+* Try to stick to immutable objects — Immutable objects are straightforward to create as materializing an object is then a matter of calling its constructor only. Also, this avoids your domain objects to be littered with setter methods that allow client code to manipulate the objects state. If you need those, prefer to make them package protected so that they can only be invoked by a limited amount of co-located types. Constructor-only materialization is up to 30% faster than properties population.
+
+* Provide an all-args constructor — Even if you cannot or don’t want to model your entities as immutable values, there’s still value in providing a constructor that takes all properties of the entity as arguments, including the mutable ones, as this allows the object mapping to skip the property population for optimal performance.
+____
+
+What's more, even though...
+
+> All Spring Data modules are inspired by the concepts of “repository”, “aggregate”, and “aggregate root” from Domain Driven Design.
+> -- Spring Data JDBC Reference Documentation, https://docs.spring.io/spring-data/jdbc/docs/2.1.7/reference/html/#jdbc.domain-driven-design
+
+...every single Spring Data JPA-based project I've ever encountered in the wild was nothing like that. Usually, they'd ignore DDD, create a repository per table, and have a complete bi-directed graph of all entities.
+
+It seems like the team behind SJD is of the same opinion:
+
+> These are possibly even more important for Spring Data JDBC, because they are, to some extent, contrary to normal practice when working with relational databases.
+> -- Spring Data JDBC Reference Documentation
+
+This so-called "normal practice" is a disastrous nightmare from design, maintenance, and performance standpoints. It can provide you with a quick temporary solution, though.
```

Comment:
```suggestion
This so-called "normal practice" is a disastrous nightmare from design, maintenance, and performance standpoints. It can provide you with a quick solution for problem at hands, though.
```